File: NickvisionMoney.Shared/Models/CurrencyConversionService.cs
using Nickvision.Aura;
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace NickvisionMoney.Shared.Models;

/// <summary>
/// A model of a result of a currency conversion
/// </summary>
public class CurrencyConversion
{
    /// <summary>
    /// The currency code of the source amount
    /// </summary>
    public string SourceCurrency { get; init; }
    /// <summary>
    /// The source amount to convert
    /// </summary>
    public decimal SourceAmount { get; init; }
    /// <summary>
    /// The currency code for the result amount
    /// </summary>
    public string ResultCurrency { get; init; }
    /// <summary>
    /// The rate of conversion from the source currency to the result currency
    /// </summary>
    public decimal ConversionRate { get; init; }

    /// <summary>
    /// The result amount
    /// </summary>
    public decimal ResultAmount => SourceAmount * ConversionRate;

    /// <summary>
    /// Constructs a CurrencyConversion
    /// </summary>
    /// <param name="sourceCurrency">The currency code of the source amount</param>
    /// <param name="sourceAmount">The source amount to convert</param>
    /// <param name="resultCurrency">The currency code for the result amount</param>
    /// <param name="conversionRate">The rate of conversion from the source currency to the result currency</param>
    public CurrencyConversion(string sourceCurrency, decimal sourceAmount, string resultCurrency, decimal conversionRate)
    {
        SourceCurrency = sourceCurrency;
        SourceAmount = sourceAmount;
        ResultCurrency = resultCurrency;
        ConversionRate = conversionRate;
    }
}

/// <summary>
/// A service for getting currency conversions
/// </summary>
public static class CurrencyConversionService
{
    private static readonly HttpClient _http;

    /// <summary>
    /// Constructs a static CurrencyConversionService
    /// </summary>
    static CurrencyConversionService()
    {
        _http = new HttpClient();
        _http.DefaultRequestHeaders.UserAgent.ParseAdd("Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/116.0");
    }

    /// <summary>
    /// Converts a sourceAmount from the sourceCurrency to the resultCurrency
    /// </summary>
    /// <param name="sourceCurrency">The currency code of the source amount</param>
    /// <param name="sourceAmount">The source amount to convert</param>
    /// <param name="resultCurrency">The currency code for the result amount</param>
    /// <returns>CurrencyConversion if successful, else null</returns>
    public static async Task<CurrencyConversion?> ConvertAsync(string sourceCurrency, decimal sourceAmount, string resultCurrency)
    {
        if (sourceCurrency == resultCurrency)
        {
            return new CurrencyConversion(sourceCurrency, sourceAmount, resultCurrency, 1);
        }
        var rates = await GetConversionRatesAsync(sourceCurrency);
        if (rates == null || !rates.ContainsKey(resultCurrency))
        {
            return null;
        }
        return new CurrencyConversion(sourceCurrency, sourceAmount, resultCurrency, rates[resultCurrency]);
    }

    /// <summary>
    /// Gets a dictionary of conversion rates from the source currency
    /// </summary>
    /// <param name="sourceCurrency">The currency code to get converting rates for</param>
    /// <returns>Dictionary&lt;string, decimal&gt; is successful, else false</returns>
    /// <remarks>This method will cache the data for the sourceCurrency on disk</remarks>
    public static async Task<Dictionary<string, decimal>?> GetConversionRatesAsync(string sourceCurrency)
    {
        var path = $"{UserDirectories.ApplicationCache}{Path.DirectorySeparatorChar}currency_{sourceCurrency}.json";
        var needsUpdate = !File.Exists(path);
        JsonDocument? json = null;
        if (!needsUpdate) //File.Exists(path)
        {
            try
            {
                json = JsonDocument.Parse(await File.ReadAllTextAsync(path));
                var seconds = json.RootElement.GetProperty("time_next_update_unix").GetInt64();
                if (DateTimeOffset.FromUnixTimeSeconds(seconds).ToLocalTime() <= DateTime.Now)
                {
                    needsUpdate = true;
                    json.Dispose();
                }
            }
            catch (Exception e)
            {
                // Couldn't get the cached rates
                Console.Error.WriteLine(e);
                needsUpdate = true;
                json?.Dispose();
            }
        }
        if (needsUpdate)
        {
            var apiUrl = $"https://open.er-api.com/v6/latest/{sourceCurrency}";
            try
            {
                var response = await _http.GetStringAsync(apiUrl);
                json = JsonDocument.Parse(response);
                if (json.RootElement.GetProperty("result").GetString() != "success")
                {
                    json.Dispose();
                    return null;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e);
                json?.Dispose();
                return null;
            }
        }
        if (json != null)
        {
            try
            {
                var ratesJson = json.RootElement.GetProperty("rates").ToString() ?? "";
                var rates = JsonSerializer.Deserialize<Dictionary<string, decimal>>(ratesJson);
                if (needsUpdate)
                {
                    try
                    {
                        await File.WriteAllTextAsync(path, json.RootElement.ToString() ?? "");
                    }
                    catch (Exception e)
                    {
                        Console.Error.WriteLine(e);
                    }
                }
                return rates;
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e);
                return null;
            }
            finally
            {
                json.Dispose();
            }
        }
        return null;
    }
}

File: libdenaro/src/models/currencyconversionservice.cpp
#include "models/currencyconversionservice.h"
#include <cstdint>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <unordered_map>
#include <json/json.h>
#include <libnick/app/aura.h>
#include <libnick/filesystem/userdirectories.h>
#include <libnick/helpers/webhelpers.h>

using namespace Nickvision::App;
using namespace Nickvision::Filesystem;

namespace Nickvision::Money::Shared::Models
{
    std::optional<CurrencyConversion> CurrencyConversionService::convert(const std::string& sourceCurrency, double sourceAmount, const std::string& resultCurrency)
    {
        if(sourceCurrency == resultCurrency)
        {
            return CurrencyConversion{ sourceCurrency, sourceAmount, resultCurrency, 1 };
        }
        const std::map<std::string, double>& rates{ getConversionRates(sourceCurrency) };
        if(rates.empty() || !rates.contains(resultCurrency))
        {
            return std::nullopt;
        }
        return CurrencyConversion{ sourceCurrency, sourceAmount, resultCurrency, rates.at(resultCurrency) };
    }

    const std::map<std::string, double>& CurrencyConversionService::getConversionRates(const std::string& sourceCurrency)
    {
        static std::unordered_map<std::string, std::map<std::string, double>> cache;
        std::filesystem::path path{ UserDirectories::getApplicationCache() / ("currency_" + sourceCurrency + ".json") };
        bool needsUpdate{ !std::filesystem::exists(path) };
        Json::Value json;
        if(std::filesystem::exists(path))
        {
            std::ifstream in{ path };
            in >> json;
            std::int64_t seconds{ json.get("time_next_update_unix", 0).asInt64() };
            std::int64_t secondsNow{ std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count() };
            if(seconds <= secondsNow)
            {
                needsUpdate = true;
                json.clear();
                cache[sourceCurrency].clear();
                Aura::getActive().getLogger().log(Logging::LogLevel::Debug, sourceCurrency + " rates on disk out-of-date.");
            }
            else if(!cache[sourceCurrency].empty())
            {
                Aura::getActive().getLogger().log(Logging::LogLevel::Debug, sourceCurrency + " rates fetched from cache.");
                return cache[sourceCurrency];
            }
        }
        if(needsUpdate)
        {
            std::string apiUrl{ "https://open.er-api.com/v6/latest/" + sourceCurrency };
            std::string response{ WebHelpers::fetchJsonString(apiUrl) };
            if(!response.empty())
            {
                Json::Reader reader;
                reader.parse(response, json, false);
                Aura::getActive().getLogger().log(Logging::LogLevel::Debug, sourceCurrency + " rates fetched from internet.");
            }
        }
        if(!json.empty())
        {
            Json::Value ratesJson{ json.get("rates", {}) };
            if(!ratesJson.empty())
            {
                std::map<std::string, double>& rates{ cache[sourceCurrency] };
                double sourceRate{ ratesJson.get(sourceCurrency, 0.0).asDouble() };
                for(const std::string& rate : ratesJson.getMemberNames())
                {
                    rates[rate] = sourceRate / ratesJson.get(rate, 0.0).asDouble();
                }
                Aura::getActive().getLogger().log(Logging::LogLevel::Debug, sourceCurrency + " rates cached.");
                if(needsUpdate)
                {
                    std::ofstream out{ path };
                    out << json;
                    Aura::getActive().getLogger().log(Logging::LogLevel::Debug, sourceCurrency + " rates saved to disk.");
                }
            }
        }
        return cache[sourceCurrency];
    }
}
